pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +321,563 @@ pub mod pallet {
+				});
+
+				ensure!(
+					info.foreign_currency == payout_foreign_currency,
+					Error::<T>::MismatchedForeignCurrency
+				);
+
+				info.total_tranche_tokens
+					.ensure_add_assign(tranche_tokens_amount)?;
+
+				Ok(())
+			})?;
+
+			T::Investment::update_redemption(
+				who,
+				investment_id,
+				T::Investment::redemption(who, investment_id)?.ensure_add(tranche_tokens_amount)?,
+			)
+		}
+
+		fn decrease_foreign_redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			tranche_tokens_amount: T::Balance,
+			payout_foreign_currency: T::CurrencyId,
+		) -> Result<(T::Balance, T::Balance), DispatchError> {
+			ForeignRedemptionInfo::<T>::mutate(&who, investment_id, |info| -> DispatchResult {
+				let info = info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+				info.total_tranche_tokens
+					.ensure_sub_assign(tranche_tokens_amount)?;
+
+				ensure!(
+					info.foreign_currency == payout_foreign_currency,
+					Error::<T>::MismatchedForeignCurrency
+				);
+
+				Ok(())
+			})?;
+
+			T::Investment::update_redemption(
+				who,
+				investment_id,
+				T::Investment::redemption(who, investment_id)?.ensure_sub(tranche_tokens_amount)?,
+			)?;
+
+			let remaining_amount = T::Investment::redemption(who, investment_id)?;
+
+			Ok((tranche_tokens_amount, remaining_amount))
+		}
+
+		fn collect_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			_foreign_payment_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::collect_investment(who.clone(), investment_id)
+		}
+
+		fn collect_foreign_redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			_foreign_payout_currency: T::CurrencyId,
+			_pool_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::collect_redemption(who.clone(), investment_id)
+		}
+
+		fn investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::investment(who, investment_id)
+		}
+
+		fn redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::redemption(who, investment_id)
+		}
+
+		fn accepted_payment_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
```

Comment:
This implementation is the same as before. Wondering why it could return `true`, even if `Investment::accepted_payout_currency()` return `false`. Why the `else` branch?

cc @wischli

Reply:
Else is bascially going to swap into `pool_currency`, while `T::Investment::accepted_payout_currency` does just check the if `currency` is equal `pool_currency`

Reply:
So, if the currency is not accepted by `T::Investment::accepted_payout_currency`, then it could be accepted for FI anyway by the else branch.